### ISSUE TITLE
[testcase notes] #27: rationale for Github pull request

### ISSUE THREAD

**Issue by dorawyy** [dorawyy/git-merge-conflicts-test#67]
# Materials from Github Help
* [Collaborating with issues and pull requests](https://help.github.com/categories/collaborating-with-issues-and-pull-requests/)
* [Creating a pull request from a fork](https://help.github.com/articles/creating-a-pull-request-from-a-fork/)
    * it mentioned that when making pull request between forks, the pull request creator can choose whether to allow the upstream repo owner make commits on the fork or not.
* [Defining the mergeability of pull requests](https://help.github.com/articles/defining-the-mergeability-of-pull-requests/)
    * [protected branch](https://help.github.com/articles/configuring-protected-branches/)
    <img width="1280" alt="screen shot 2017-11-10 at 10 56 57 am" src="https://user-images.githubusercontent.com/16771854/32673941-eda889e4-c605-11e7-8c3e-61ca0484ec91.png">

    * [status check](https://help.github.com/articles/about-required-status-checks/)
        * developers can set up protected branches, which means every commit pushed to that the protected branch ( base branch in pull requests) has to pass all status checks
        * [Types of required status checks](https://help.github.com/articles/types-of-required-status-checks/) : strict/loose/disabled
        * [Enabling required status checks](https://help.github.com/articles/enabling-required-status-checks/): make use of the [Status API](https://developer.github.com/v3/repos/statuses/)
        * [About branch restrictions](https://help.github.com/articles/about-branch-restrictions/)
        * [About required reviews for pull requests](https://help.github.com/articles/about-required-reviews-for-pull-requests/)

**Comment by dorawyy** [comment]
# Findings
* [Changing the base branch of a pull request](https://help.github.com/articles/changing-the-base-branch-of-a-pull-request/)
    * After a pull request is made, the `base branch` can still be changed, but the `head branch` cannot be changed
![1510343221050](https://user-images.githubusercontent.com/16771854/32675744-e1adecb8-c60c-11e7-88ce-602892037647.jpg)
* [Committing changes to a pull request branch created from a fork](https://help.github.com/articles/committing-changes-to-a-pull-request-branch-created-from-a-fork/)
    * Generally speaking, as the owner of the upstream repo, you don't have access to the pull request branch (from the fork); However, the person who made the pull request can grant you access to commit on his/her pull request branch
    * Also, you can make a fork of the forked repo, then make commit on your fork of the forked repo, then pull request to merge commits to the forked repo.

**Comment by dorawyy** [comment]
# Addressing merge conflicts in pull request
### [About merge conflicts](https://help.github.com/articles/about-merge-conflicts/)
* You must resolve the conflict before you can merge the branches.
* You cannot merge your pull request branch on GitHub until you resolve the merge conflict
* To resolve merge conflicts, you have two options
    * [merge on Github](https://help.github.com/articles/resolving-a-merge-conflict-on-github/) if the conflict is a [competing line conflict](https://help.github.com/articles/resolving-a-merge-conflict-using-the-command-line/#competing-line-change-merge-conflicts)
    * [merge locally on command line](https://help.github.com/articles/resolving-a-merge-conflict-using-the-command-line/#competing-line-change-merge-conflicts) for any type of conflicts, namely (1)competing line conflict(aka, content modification conflict); (2) modification/deletion conflict.

### [Resolving a merge conflict on GitHub](https://help.github.com/articles/resolving-a-merge-conflict-on-github/)
* Only competing line conflicts can be resolved on Github, all other merge conflicts have to be resolved locally via command line ( -- for pull request proposer)
* When all conflicts are resolved, clock *`Commit merge`* in Github merge tool, **this will merge the entire base branch into your head branch (pull request branch) **
* Then, the pull request can be further merged.

### []()

**Comment by dorawyy** [comment]
# [Merging a pull request](https://help.github.com/articles/merging-a-pull-request/)

**Anyone with push access to the repo can complete the merge.**

If the pull request does not have any merge conflicts, you can merge it on Github. If the pull request does have merge conflicts, or if you'd like to rest the changes before merging, you can [check out the pull request locally](https://help.github.com/articles/checking-out-pull-requests-locally/) and merge it using the command line.

If you decide you don't want the changes in a topic branch to be merged to the upstream branch, you can [close the pull request](https://help.github.com/articles/closing-a-pull-request/) without merging.

![screen shot 2017-11-10 at 5 47 58 pm](https://user-images.githubusercontent.com/16771854/32685058-57e51ad2-c63f-11e7-8acb-ac2e838f961e.png)


## [Checking out pull requests locally](https://help.github.com/articles/checking-out-pull-requests-locally/) 
* Basically, follow the command line instruction here (is the instruction of option1, merge pull request)
<img width="858" alt="wx20171110-175143 2x" src="https://user-images.githubusercontent.com/16771854/32685091-e2dc246e-c63f-11e7-9e7f-b5295e479b4e.png">
(as can be seen from the screenshot above, forbids fast-forward merging with `--no-ff` param)

**Comment by dorawyy** [comment]
# [Merging a pull request on GitHub](https://help.github.com/articles/merging-a-pull-request/#merging-a-pull-request-on-github)
* Option1: retaining all the commits in a feature branch;
* Option2: squashing all commits into a single commit
* Option3: rebasing individual commits from the head branch onto the base branch.


## [Merge all of the commits into the base branch](https://help.github.com/articles/about-pull-request-merges/)
* To merge pull requests, you must have write permissions in the repository.
* Actually do a n-way merge with `--no-ff` param:
```
git checkout BaseBranch
git merge HeadBranch --no-ff
```
<img width="862" alt="screen shot 2017-11-11 at 12 35 55 am" src="https://user-images.githubusercontent.com/16771854/32687753-4ff7318c-c678-11e7-844b-73bee40ca65e.png">

## [Squash and merge your pull request commits](https://help.github.com/articles/about-pull-request-merges/#squash-and-merge-your-pull-request-commits)
* The pull request's commits are squashed into a single commit.
* Pull requests with squashed commits are merged using the fast-forward option.
* Requirements:
    * you must have write permissions in the repository, 
    * and the repository must allow squash merging.
<img width="870" alt="screen shot 2017-11-11 at 12 44 09 am" src="https://user-images.githubusercontent.com/16771854/32687816-74bea026-c679-11e7-98d5-8d1a2e7531b5.png">

## [Rebase and merge your pull request commits](https://help.github.com/articles/about-pull-request-merges/#rebase-and-merge-your-pull-request-commits)
* all commits from the head branch (or pull request branch) are added onto the base branch individually without a merge commit
* Requirement:
    * you must have write permissions in the repository, 
    * and the repository must allow rebase merging
* The rebase and merge behavior on GitHub deviates slightly from git rebase.
    * **Rebase and merge on GitHub will always update the committer information and create new commit SHAs;** 
    * whereas git rebase outside of GitHub does not change the committer information when the rebase happens on top of an ancestor commit. 
* **You aren't able to automatically rebase and merge on GitHub when:**
    * **The pull request has merge conflicts.**
    * **Rebasing the commits from the base branch into the head branch runs into conflicts.**
    * **Rebasing the commits is considered "unsafe," such as when a rebase is possible without merge conflicts but would produce a different result than a merge would.**


* If you still want to rebase the commits but can't rebase and merge automatically on GitHub you must:
    * Rebase the topic branch (or head branch) onto the base branch locally on the command line
    * Resolve any merge conflicts on the command line.
    * Force-push the rebased commits to the pull request's topic branch (or remote head branch).

**Comment by dorawyy** [comment]
# [Comparing commits across time (Compare View)](https://help.github.com/articles/comparing-commits-across-time/)

* Every repository contains a **Compare view**, which allows you to compare the state of your repository across branches, tags, commits, time periods, and more. The compare view provides you with the same diff tooling that the Pull Request view does.

* To get to the compare view, append `/compare` to your repository's path.

* **_Link to the compare view of the current repo:_** 
[https://github.com/dorawyy/git-merge-conflicts-test/compare](https://github.com/dorawyy/git-merge-conflicts-test/compare)

* Every repository's Compare view contains two drop down menus: `base` and `compare`.
<img width="1042" alt="screen shot 2017-11-11 at 12 31 05 pm" src="https://user-images.githubusercontent.com/16771854/32693246-6b95c822-c6dc-11e7-9003-badb2d823699.png">

* **`base` should be considered the starting point of your comparison, and `compare` is the endpoint.**
* During a comparison, you can always change your `base` and `compare` points by clicking on _Edit_.

## Comparing branches
* The most common use of Compare is to compare branches, such as when you're starting a new Pull Request. **You'll always be taken to the branch comparison view when starting a new Pull Request.**
* To compare branches, you can select a branch name from the `compare` drop down menu at the top of the page.
* Here's an [example of a comparison between two branches](https://github.com/octocat/linguist/compare/master...octocat:an-example-comparison-for-docs).

## Comparing tags
* Similarly, you can compare across tags made for project releases. Comparing against tags is a great way to assemble release notes between different versions of your project.
* Instead of typing a branch name, type the name of your tag in the `compare` drop down menu.
* Here's an [example of a comparison between two tags](https://github.com/octocat/linguist/compare/v2.2.0...octocat:v2.3.3).

## Comparing commits
* You can also compare two arbitrary commits in your repository. Comparisons between commits are made by providing either the full SHA hash or the short seven-character code.
* Here's an [example of a comparison between two commits](https://github.com/octocat/linguist/compare/e84204a27468503bcdc47f80c0f0d7a3c2341c95...octocat:abfe89d8ff14416c391b3845e5a4c5ffa192b296).

## Comparing across forks
* You can compare your base repository and any forked repository. This is the view that's presented when a user performs a Pull Request to a project.
* To compare branches on different repositories, preface the branch names with user names. For example, by specifying `octocat:master` for base and `octo-org:master` for compare, you can compare the master branch of the repositories respectively owned by `octocat` and `octo-org`.
* Here's an [example of a comparison between two repositories](https://github.com/octocat/linguist/compare/master...octo-org:master).

## Comparisons across time
* Comparisons can be created for arbitrary time periods, like one month or two weeks. To define a time period, type the branch name, followed by a `@`, and then the date wrapped between a `{ }` notation. For example, typing `master@{2weeks}` into the `base` dropdown menu compares the current `master` branch against the `master` branch as it was two weeks prior.
* Here's an [example of a comparison between two time periods](https://github.com/octo-org/wikimania/compare/master@%7B2weeks%7D...master).
* You can also specify a specific date to compare against. Date formatting must follow the [ISO8601](https://en.wikipedia.org/wiki/ISO_8601) standard, which is `YYYY-MM-DD` (year-month-day).
* Here's an [example comparing a branch from 2016-07-22 against that same branch several days later](https://github.com/octo-org/wikimania/compare/master@%7B07-22-16%7D...master@%7B08-04-16%7D).

## Comparisons across commits
* As a shortcut, Git uses the `^` notation to mean "one commit prior."
* You can use this notation to compare a single commit or branch against its immediate predecessors. For example, `96d29b7^^^^^` indicates five commits prior to `96d29b7`, because there are five `^` marks. Typing `96d29b7^^^^^` in the `base` branch and `96d29b7` in the `compare` branch compares the five commits made before `96d29b7` with the `96d29b7` commit.

Here's an [example of a comparison using the ^ notation](https://github.com/octocat/linguist/compare/octocat:96d29b7%5E%5E%5E%5E%5E...octocat:96d29b7).

**Comment by dorawyy** [comment]
# By default, can original repo's owner push to fork (both public)?
**Answer:** 
* No, if the repo is a public repo, then by default, original repo's collaborators don't have push access to the fork repo. 

## Here are the details:
### 1. Original repo's colalborators can propose commits to the fork, all commits they proposed will go via pull requests
<img width="832" alt="wx20171117-095341 2x" src="https://user-images.githubusercontent.com/16771854/32961579-f0aed580-cb7d-11e7-98a9-4f6a10787119.png">

![1510941456960](https://user-images.githubusercontent.com/16771854/32961597-07ec5e8e-cb7e-11e7-8488-d9e4f729100c.jpg)

### 2. Then transferring to the compare view (comparing the patch branch on original repo and the branch you want to commit to of the fork repo
<img width="917" alt="wx20171117-100332 2x" src="https://user-images.githubusercontent.com/16771854/32961753-a042603e-cb7e-11e7-9c28-a81a6687b28d.png">


### 3.  All commits will be made in the committer's own repo (original repo in this testcase), via checking out a new patch branch
<img width="970" alt="wx20171117-100615 2x" src="https://user-images.githubusercontent.com/16771854/32961865-244c714e-cb7f-11e7-977d-caa94d7a7884.png">
<img width="907" alt="wx20171117-100654 2x" src="https://user-images.githubusercontent.com/16771854/32961866-246344f0-cb7f-11e7-892a-24ca5b7fa743.png">
<img width="1000" alt="wx20171117-100723 2x" src="https://user-images.githubusercontent.com/16771854/32961867-24753db8-cb7f-11e7-89f2-9f6a1000e38f.png">

**Comment by dorawyy** [comment]
# By default, the pull request would grant original repo's owner push access to the fork repo branch

![wx20171117-101947](https://user-images.githubusercontent.com/16771854/32962379-159f9264-cb81-11e7-9962-25d94a4afec7.png)

The developer needs to unckeck this
